### DISCUSSION TITLE
high level explanation of RustDesk's end-to-end encryption?

### DISCUSSION THREAD

**Discussion by kolAflash** [rustdesk/rustdesk#2239] [Q&A] [answered]
Can someone give me a high level explanation of the encryption RustDesk does?
[rustdesk.com](https://rustdesk.com/) states: _"End-to-end encryption."_
So when doing end to end encryption, I usually expect one of the following solutions to verify the others identity.

&nbsp;

**PKI**
Either [certificate authorities](https://en.wikipedia.org/wiki/Certificate_authority) for a public key infrastructure. And each client has to get a certificate for it's "identity" (compare: [X.509](https://en.wikipedia.org/wiki/X.509) certificates email addresses to use [S/MIME encryption](https://en.wikipedia.org/wiki/S/MIME)).

**Fingerprints**
Or the alternative I'd expect is some manual public key fingerprint comparison. Either by scanning QR codes, like some mobile messengers do it ([Signal](https://en.wikipedia.org/wiki/Signal_(software)), [Threema](https://en.wikipedia.org/wiki/Threema), ...). Or by manually comparing fingerprints, like people are used to do it when using [PGP](https://en.wikipedia.org/wiki/Pretty_Good_Privacy) or [SSH](https://en.wikipedia.org/wiki/Secure_Shell).
_(I'd clearly prefer that one)_

&nbsp;

I couldn't find neither in RustDesk. Or did I miss a secret menu to show fingerprints to compare them or install my X.509 certificate?

I really like RustDesk. And I'd really welcome a way to compare fingerprints, including some [TOFU](https://en.wikipedia.org/wiki/Trust_on_first_use) mechanism to remember identities I've verified before. Maybe RustDesk can even incooperate PGP, to the identity can simply be verified by the PGP key.

In some way this relates to the question which has been asked here before: https://github.com/rustdesk/rustdesk/issues/223
_"How does this end-to-end encryption prevent man-in-the-middle attacks?"_
Or in other words: How does RustDesk [authenticate identities](https://en.wikipedia.org/wiki/Man-in-the-middle_attack#Authentication) to establish a secure end-to-end encryption?

**Comment by djex** [comment]
I do not know the details on how the encryption is implemented however I did find that it uses the NaCl encryption library and uses crypto_secretbox_xsalsa20poly1305 for encryption.

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk/issues/63#issuecomment-855530293

> crypto_sign_ed25519 for signature creation and verification
crypto_box_curve25519xsalsa20poly1305 for asymmetric cryptography
crypto_secretbox_xsalsa20poly1305 for symmetric cryptography

**Comment by AlexZigante** [comment]
I would love a detail explanation too.
Mainly does the end-to-end mean between machines or is it just encrypted machine to server then decrypted on server and re-encrypted to transmit to second machine??

**Reply by rustdesk** [reply]
Between machines

**Comment by rustdesk** [comment, marked as answer]
> fingerprints

We do have this, but not displayed friendly on 

controlled side 

> @21pages please check where anydesk show it, I guess somewhere in settings or about page

and remote menubar 

> @21pages  should be shown aside the secured connection tip when the mouse hover on tab of the remote window (where for mobile?).

yet. @21pages add this in our GUI please. For us, the fingerprint is the public key of controlled side.

https://github.com/rustdesk/rustdesk/blob/bdd314856f798080288b01e1b0f4a4e6b3f89ecb/src/client.rs#L476

https://github.com/rustdesk/rustdesk/blob/7a818c54f81de7965ae3ad53c0e36db4aa148d7e/src/server.rs#L144

**Comment by 21pages** [comment]
anydesk about page
![c423c089e14f8260828c933c9a7e260](https://user-images.githubusercontent.com/14891774/232757232-ab95504c-c596-4c80-8660-8cd960e53db1.png)

**Reply by rustdesk** [reply]
Let's follow AnyDesk.

**Comment by rustdesk** [comment]
https://github.com/rustdesk/rustdesk/pull/4098

> ![1681890618299](https://user-images.githubusercontent.com/14891774/233006268-a5767819-bb79-4cfa-bdd2-dbdc8eecb5a3.png)
![a5ef2107fda0291e529570bb47be4a0](https://user-images.githubusercontent.com/14891774/232997649-95848efd-00d8-4b9a-bc3a-0ca6f814e414.jpg)

**Reply by kolAflash** [reply]
Thanks a lot!

I know many decent IT security and network administrator guys who will honor this.
